### PR TITLE
feat: add support for Substrait ExtendedExpression

### PR DESCRIPTION
**Pull request by westonpace**
## Which issue does this PR close?

This partially addresses #7687 (though the issue talks about a python component which is not yet included here)

## Rationale for this change

Described in issue.

## What changes are included in this PR?

Adds serialization to and from Substrait's top-level ExtendedExpression message

## Are these changes tested?

Yes, new unit tests are added.

## Are there any user-facing changes?

New methods.  No changes to existing behavior.

### CODE REVIEW

**Comment by alamb** [comment]
@vbarua  / @Blizzara  do either of you have time to review this PR?

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 1]
```diff
@@ -251,6 +252,147 @@ pub async fn from_substrait_plan(
...
+    /// The expressions
+    ///
+    /// Each item contains an expression and the field that defines the expected nullability and name of the expr's output
+    pub exprs: Vec<(Expr, Field)>,
+}
+
+// Substrait fields (DataType, nullable) typically don't have names.  In the few cases they do
+// those names are provided as Vec<String>.  This method attaches names to a (DataType, nullable)
+// pair to create a Field
+fn field_from_type_and_names<S: Into<String>, T: Iterator<Item = S>>(
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:273]: This feels similar to `from_substrait_type`, could we reuse that somehow?

Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/consumer.rs:273]: Or actually maybe `make_renamed_schema` is what you want?

Comment (westonpace) [reply datafusion/substrait/src/logical_plan/consumer.rs]: Good catch.  I found the existing method in the producer but missed it in the consumer.  I've updated the PR to use `make_renamed_schema`'s helper method (which I've renamed as `rename_field`)

**Review by alamb — APPROVED** [review]
I reviewed this code and found it well commented and easy to follow

While I am not an expert on this particular part of substrait, I think it is a nice addition to DataFusion and one we can keep iterating on if needed

Thank you for the contribution @westonpace  and the review @Blizzara

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 2]
```diff
@@ -251,6 +252,81 @@ pub async fn from_substrait_plan(
...
+            }
+            None => {
+                plan_err!("required property `expr_type` missing from Substrait ExpressionReference message")
+            }
+        }?;
+        let expr =
+            from_substrait_rex(ctx, scalar_expr, &input_schema, &extensions).await?;
+        let (output_type, expected_nullability) =
+            expr.data_type_and_nullable(&input_schema)?;
+        let output_field = Field::new("", output_type, expected_nullability);
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:312]: You could have just picked the first name here, then started names_idx from 1 instead 😅 I'm overall not a super fan of the "rename_self" option but I see why you did it, I guess it's alright.

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 3]
```diff
@@ -334,6 +410,68 @@ fn rename_expressions(
         .collect()
 }
 
+fn rename_field(
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/consumer.rs:413]: nice, and thanks for adding the comments!

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 4]
```diff
@@ -2461,4 +2508,101 @@ mod test {
 
         Ok(())
     }
+
+    #[test]
+    fn named_struct_names() -> Result<()> {
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:2513]: thanks!

**Review by Blizzara — APPROVED** [review]
Just did another round, LGTM! 

As a aside, I feel like we should look into splitting the consumer.rs and producer.rs into couple files, as they're growing quite large...

**Comment by alamb** [comment]
> Just did another round, LGTM!
> 
> As a aside, I feel like we should look into splitting the consumer.rs and producer.rs into couple files, as they're growing quite large...

Thanks @Blizzara  🙏 

Let's do the other suggestions as a follow on PR (I like the idea of splitting consumer/producer)

Thanks again @westonpace

**Comment by westonpace** [comment]
> As a aside, I feel like we should look into splitting the consumer.rs and producer.rs into couple files, as they're growing quite large...

@Blizzara agreed.  One simple split for now could be to split the expressions/literals/types/schema from the relations.  At some distant point (not in my current plans) I could imagine the crate itself splitting to create a lightweight expression manipulation library that depends only on `datafusion-expr`.

**Comment by alamb** [comment]
I have found the mechanical "split the code into different files" is always a good first step towards anything more substantial

**Pull request merged**